pgmpy/inference/dbn_inference.py

```diff
@@ -0,0 +1,207 @@
+        self.interface_nodes_1 = model.get_interface_nodes(1)
+
+        start_markov_model = self.start_bayesian_model.to_markov_model()
+        one_and_half_markov_model = self.one_and_half_model.to_markov_model()
+
+        combinations_slice_0 = list(itertools.combinations(model.get_interface_nodes(0), 2))
+        combinations_slice_1 = list(itertools.combinations(model.get_interface_nodes(1), 2))
+
+        start_markov_model.add_edges_from(combinations_slice_0)
+        one_and_half_markov_model.add_edges_from(combinations_slice_0 + combinations_slice_1)
+
+        self.one_and_half_junction_tree = one_and_half_markov_model.to_junction_tree()
+        self.start_junction_tree = start_markov_model.to_junction_tree()
+
+        self.start_interface_clique = self._get_clique(self.start_junction_tree, self.interface_nodes_0)
+        self.in_clique = self._get_clique(self.one_and_half_junction_tree, self.interface_nodes_0)
+        self.out_clique = self._get_clique(self.one_and_half_junction_tree, self.interface_nodes_1)
+
+    def _shift_nodes(self, nodes, time):
+        """
+        shifting the nodes to a certain required timeslice.
+        Parameters
+        ----------
+        shift: int
+            shifting the evidence corresponding to the given time slice.
+        """
+        return [(node[0], time) for node in nodes]
+
+    def _get_clique(self, junction_tree, nodes):
+        """
+        extracting the cliques from the junction tree which are a subset of
+        the given nodes.
+        Parameters
+        ----------
+        junction_tree: Junction tree
+            from which the nodes are to be extracted.
+        nodes: iterable container
+            A container of nodes (list, dict, set, etc.).
+        """
+        return [clique for clique in junction_tree.nodes() if set(nodes).issubset(clique)][0]
+
+    def _get_evidence(self, evidence_dict, time, shift):
+        """
+        getting the evidence belonging to a particular timeslice.
+        Parameters
+        ----------
+        evidence: dict
+            a dict key, value pair as {var: state_of_var_observed}
+            None if no evidence
+        time: int
+            the evidence corresponding to the time slice
+        shift: int
+            shifting the evidence corresponding to the given time slice.
+        """
+        if evidence_dict:
+            return {(node[0], shift): evidence_dict[node] for node in evidence_dict if node[1] == time}
+        else:
+            return None
+
+    def _marginalize_factor(self, nodes, factor):
+        """
+        marginalizing the factor selectively for a set of variables.
+        Parameters
+        ----------
+        nodes: iterable container
+            A container of nodes (list, dict, set, etc.).
+        factor: factor
+            factor which is to be marginalized.
+        """
+        marginalizing_nodes = list(set(factor.scope()).difference(nodes))
```

Comment:
@palashahuja No need to convert it to list. `marginalize` should work with sets also